samples/msal-node-samples/auth-code-distributed-cache/src/AuthProvider.ts

```diff
@@ -0,0 +1,192 @@
+        tokenResponse = await msalInstance.acquireTokenSilent(silentRequest);
+        performance.mark("acquireTokenSilent-end");
+        performance.measure("acquireTokenSilent", "acquireTokenSilent-start", "acquireTokenSilent-end");
+
+        return tokenResponse;
+    }
+
+    private static async getMetadata(msalConfig: Configuration, cacheClient: RedisClientType): Promise<Configuration> {
+        const msalConfigWithMetadata = msalConfig;
+        const clientId = msalConfigWithMetadata.auth.clientId;
+        const tenantId = msalConfigWithMetadata.auth.authority!.split("/").pop()!;
+
+        try {
+            let [cloudDiscoveryMetadata, authorityMetadata] = await Promise.all([
+                cacheClient.get(`${clientId}.${tenantId}.discovery-metadata`),
+                cacheClient.get(`${clientId}.${tenantId}.authority-metadata`)
+            ]);
+
+            if (!cloudDiscoveryMetadata || !authorityMetadata) {
+                [cloudDiscoveryMetadata, authorityMetadata] = await Promise.all([
+                    AuthProvider.fetchCloudDiscoveryMetadata(tenantId),
+                    AuthProvider.fetchOIDCMetadata(tenantId)
+                ]);
+
+                if (cloudDiscoveryMetadata && authorityMetadata) {
+                    await cacheClient.set(`${clientId}.${tenantId}.discovery-metadata`, JSON.stringify(cloudDiscoveryMetadata));
+                    await cacheClient.set(`${clientId}.${tenantId}.authority-metadata`, JSON.stringify(authorityMetadata));
+                }
+            }
+
+            msalConfigWithMetadata.auth.cloudDiscoveryMetadata = typeof cloudDiscoveryMetadata === 'string' ? cloudDiscoveryMetadata : JSON.stringify(cloudDiscoveryMetadata);
+            msalConfigWithMetadata.auth.authorityMetadata = typeof authorityMetadata === 'string' ? authorityMetadata : JSON.stringify(authorityMetadata);
+        } catch (error) {
+            console.log(error);
+        }
+
+        return msalConfigWithMetadata;
+    }
+
+    private static async fetchCloudDiscoveryMetadata(tenantId: string): Promise<any> {
+        const endpoint = 'https://login.microsoftonline.com/common/discovery/instance';
+
+        try {
+            const response = await AxiosHelper.callDownstreamApi(endpoint, undefined, {
+                'api-version': '1.1',
+                'authorization_endpoint': `https://login.microsoftonline.com/${tenantId}/oauth2/v2.0/authorize`
+            });
+
+            return response;
+        } catch (error) {
+            console.log(error);
+        }
+    }
+
+    private static async fetchOIDCMetadata(tenantId: string): Promise<any> {
+        const endpoint = `https://login.microsoftonline.com/${tenantId}/v2.0/.well-known/openid-configuration`;
+
+        try {
+            const response = await AxiosHelper.callDownstreamApi(endpoint);
+            return response;
+        } catch (error) {
+            console.log(error);
+        }
```

Comment:
Does there need to be a return here? What if there's an error?

Reply:
Agreed, probably a rethrow.

Reply:
I'm thinking if this throws, just swallow it and let msal fetch the metadata itself. It would degrade performance but would not crash the app. You think we should throw early?